dali/python/nvidia/dali/ops.py

```diff
@@ -844,12 +848,45 @@ def _load_arithm_ops():
 def _choose_device(inputs):
     for input in inputs:
         if isinstance(input, (tuple, list)):
-            if any(inp.device == "gpu" for inp in input):
+            if any(getattr(inp, "device", None) == "gpu" for inp in input):
                 return "gpu"
-        elif input.device == "gpu":
+        elif getattr(input, "device", None) == "gpu":
             return "gpu"
     return "cpu"
 
+def _preprocess_inputs(inputs, op_name, device, schema=None):
+    if isinstance(inputs, tuple):
+        inputs = list(inputs)
+
+    def is_input(x):
+        if isinstance(x, (_DataNode, nvidia.dali.types.ScalarConstant)):
+            return True
+        return isinstance(x, (list)) and \
+                any(isinstance(y, _DataNode) for y in x) and \
+                all(isinstance(y, (_DataNode, nvidia.dali.types.ScalarConstant)) for y in x)
+
+    for idx, inp in enumerate(inputs):
+        input_device = schema.GetInputDevice(idx) or device if schema else device
+        if not is_input(inp):
+            if not isinstance(inp, nvidia.dali.types.ScalarConstant):
+                try:
+                    inp = nvidia.dali.types.Constant(inp, device=input_device)
+                except Exception as ex:
+                    raise TypeError("""when calling operator {0}:
+Input {1} is neither a DALI `DataNode` nor a list of data nodes but `{2}`.
+Attempt to convert it to a constant node failed."""
+                    .format(op_name, idx, type(inp).__name__)) from ex
+
+            if not isinstance(inp, _DataNode):
+                inp = nvidia.dali.ops._instantiate_constant_node(input_device, inp)
+
+# Should we?
```

Comment:
I guess we could. Mixed feelings. There's a value in writing `.gpu()` since it is a reminder that this is the point in the pipeline where your data gets transferred to the gpu.

Reply:
I know. On the other hand, we already silently do it in arithmetic ops.

Reply:
Maybe we should introduce a `strict` mode where it is not possible, and this `easy` where this can happen under the hood.

Reply:
I'll remove it for now.